flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+
+		final Optional<KubernetesConfigMap> optional = kubeClient.getConfigMap(configMapName);
+		if (optional.isPresent()) {
+			final KubernetesConfigMap configMap = optional.get();
+			if (configMap.getData().containsKey(key)) {
+				return deserializeObject(configMap.getData().get(key));
+			} else {
+				throw new StateHandleStore.NotExistException(
+					"Could not find " + key + " in ConfigMap " + configMapName);
+			}
+		} else {
+			throw configMapNotExistSupplier.get();
+		}
+	}
+
+	/**
+	 * Gets all available state handles sorted by key from Kubernetes.
+	 *
+	 * @return All state handles from ConfigMap.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<Tuple2<RetrievableStateHandle<T>, String>> getAll() {
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(
+				configMap -> {
+					final List<Tuple2<RetrievableStateHandle<T>, String>> stateHandles = new ArrayList<>();
+					configMap.getData().entrySet().stream()
+						.filter(entry -> filter.test(entry.getKey()))
+						.sorted(Comparator.comparing(Map.Entry::getKey))
+						.forEach(entry -> {
+							try {
+								stateHandles.add(new Tuple2(deserializeObject(entry.getValue()), entry.getKey()));
+							} catch (Exception e) {
+								LOG.warn("ConfigMap {} contained corrupted data. Ignoring the key {}.",
+									configMapName, entry.getKey());
+							}
+						});
+					return stateHandles;
+				})
+			.orElse(Collections.emptyList());
+	}
+
+	/**
+	 * Return a list of all valid keys for state handles.
+	 *
+	 * @return List of valid state handle keys in Kubernetes ConfigMap
+	 *
+	 * @throws Exception if get state handle names from ConfigMap failed.
+	 */
+	@Override
+	public Collection<String> getAllNames() throws Exception {
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(configMap -> configMap.getData().keySet().stream()
+				.sorted(Comparator.comparing(e -> e))
+				.filter(filter)
+				.collect(Collectors.toList()))
+			.orElseThrow(configMapNotExistSupplier);
+	}
+
+	/**
+	 * Remove the key in state config map. As well as the state on external storage will be removed.
+	 * It returns the {@link RetrievableStateHandle} stored under the given state node if any.
+	 *
+	 * @param key Key to be removed from ConfigMap
+	 *
+	 * @return True if the state handle is removed successfully
+	 *
+	 * @throws Exception if removing the key or discarding the state failed
+	 */
+	@Override
+	public boolean remove(String key) throws Exception {
+		checkNotNull(key, "Key in ConfigMap.");
+		final AtomicReference<RetrievableStateHandle<T>> stateHandleRefer = new AtomicReference<>();
+
+		try {
+			return kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (KubernetesLeaderElector.hasLeadership(configMap, lockIdentity)) {
+						final String content = configMap.getData().remove(key);
+						if (content != null) {
+							try {
+								stateHandleRefer.set(deserializeObject(content));
```

Comment:
Just as a food for thought. Somehow the `kubeClient.checkAndUpdateConfigMap` does not provide a rich enough interface to also return a value from it. If `CompletableFuture<UpdateResult> checkAndUpdateConfigMap()` with `UpdateResult` being able to contain a payload like `RetrievableStateHandle<T>`, then one would not have to use the `AtomicReference` which is somewhat working around the API which is not expressive enough. Of course, the `function` argument to the `checkAndUpdateConfigMap` method would have to be pimped as well in order to return the updated configMap value and the return value of the method itself.

Reply:
I will try to think more about this and do it later. Maybe after all the codes finished, we could know what is the best pattern for this interface.